lib/SILOptimizer/Differentiation/LinearMapInfo.cpp

```diff
@@ -392,44 +323,52 @@ void LinearMapInfo::generateDifferentiationDataStructures(
     traceEnumFieldName = "predecessor";
     break;
   }
+
   for (auto &origBB : *original) {
     auto *traceEnum =
-        createBranchingTraceDecl(&origBB, derivativeFnGenSig, loopInfo);
+        createBranchingTraceDecl(&origBB, derivativeFnGenSig);
     branchingTraceDecls.insert({&origBB, traceEnum});
-    if (origBB.isEntry())
-      continue;
-    // Add branching trace enum field to corresponding linear map struct.
-    auto *linearMapStruct = getLinearMapStruct(&origBB);
-    auto *traceEnumField = addVarDecl(
-        linearMapStruct, astCtx.getIdentifier(traceEnumFieldName).str(),
-        traceEnum->getDeclaredInterfaceType());
-    linearMapStructEnumFields.insert({linearMapStruct, traceEnumField});
   }
 
-  // Do not add linear map fields for semantic member accessors, which have
-  // special-case pullback generation. Linear map structs should be empty.
-  if (isSemanticMemberAccessor(original))
-    return;
-
-  // Add linear map fields to the linear map structs.
+  // Add linear map fields to the linear map tuples.
   for (auto &origBB : *original) {
-    for (auto &inst : origBB) {
-      if (auto *ai = dyn_cast<ApplyInst>(&inst)) {
-        // Add linear map field to struct for active `apply` instructions.
-        // Skip array literal intrinsic applications since array literal
-        // initialization is linear and handled separately.
-        if (!shouldDifferentiateApplySite(ai) ||
-            ArraySemanticsCall(ai, semantics::ARRAY_UNINITIALIZED_INTRINSIC))
-          continue;
-        if (ArraySemanticsCall(ai, semantics::ARRAY_FINALIZE_INTRINSIC))
-          continue;
-        LLVM_DEBUG(getADDebugStream()
-                   << "Adding linear map struct field for " << *ai);
-        addLinearMapToStruct(context, ai);
+    SmallVector<TupleTypeElt, 4> linearTupleTypes;
+    if (!origBB.isEntry()) {
+      CanType traceEnumType = getBranchingTraceEnumLoweredType(&origBB).getASTType();
+      linearTupleTypes.emplace_back(traceEnumType,
+                                    astCtx.getIdentifier(traceEnumFieldName));
```

Comment:
... wait I don't think SIL tuples are allowed to have labels, as it's no longer a canonical type? Am I misremembering this?

Reply:
Well, it seems it does not cause any problems anywhere, the identifier is just stored in the type and that's it. I double checked – the generated SIL parses as well.

On the other hand – having named field for branch trace enum helps debugging as we're having internal contract with indices :)

Reply:
Ok.  Something unrelated: would it be better if we made the branching trace enum be the _last_ element? Might help with alignment in the tuple.

Reply:
Frankly speaking, I do not know :) I wanted to replicate the struct layout to minimize possible unknown fallback. This is certainly worth exploring if e.g. this might improve the codegen, etc.